File: VB/Example/TreeListCheckHelper.vb
Imports Microsoft.VisualBasic
Imports DevExpress.XtraTreeList
Imports DevExpress.XtraTreeList.Nodes
Imports System
Imports System.Linq
Imports System.Windows.Forms

Namespace Example
	Friend Class TreeListCheckHelper
		Public Sub New(ByVal treeList As TreeList)
			treeList.OptionsSelection.MultiSelect = True
			treeList.OptionsView.ShowCheckBoxes = True
			treeList.OptionsBehavior.AllowIndeterminateCheckState = False
			AddHandler treeList.BeforeCheckNode, AddressOf treeList1_BeforeCheckNode
		End Sub

		Private Function AnyChildIsChecked(ByVal node As TreeListNode) As Boolean
			For Each childNode As TreeListNode In node.Nodes
				If childNode.CheckState = CheckState.Checked Then
					Return True
				End If
			Next childNode
			Return False
		End Function
		Private Sub treeList1_BeforeCheckNode(ByVal sender As Object, ByVal e As CheckNodeEventArgs)
			Dim node As TreeListNode = e.Node
			If node.Checked Then
				node.UncheckAll()
			Else
				node.CheckAll()
			End If
			Do While node.ParentNode IsNot Nothing
				node = node.ParentNode
				node.CheckState = If(AnyChildIsChecked(node), CheckState.Checked, CheckState.Unchecked)
			Loop
		End Sub
	End Class
End Namespace

File: CS/Example/TreeListCheckHelper.cs
using DevExpress.XtraTreeList;
using DevExpress.XtraTreeList.Nodes;
using System;
using System.Linq;
using System.Windows.Forms;

namespace Example
{
    class TreeListCheckHelper
    {
        public TreeListCheckHelper(TreeList treeList)
        {
            treeList.OptionsSelection.MultiSelect = true;
            treeList.OptionsView.ShowCheckBoxes = true;
            treeList.OptionsBehavior.AllowIndeterminateCheckState = false;
            treeList.BeforeCheckNode += treeList1_BeforeCheckNode;
        }

        private bool AnyChildIsChecked(TreeListNode node)
        {
            foreach(TreeListNode childNode in node.Nodes)
                if(childNode.CheckState == CheckState.Checked)
                    return true;
            return false;
        }
        private void treeList1_BeforeCheckNode(object sender, CheckNodeEventArgs e)
        {
            TreeListNode node = e.Node;
            if(node.Checked)
                node.UncheckAll();
            else node.CheckAll();
            while(node.ParentNode != null)
            {
                node = node.ParentNode;
                node.CheckState = AnyChildIsChecked(node) ? CheckState.Checked : CheckState.Unchecked;
            }
        }
    }
}